grades/api_test.py

```diff
@@ -0,0 +1,237 @@
+        assert grade2_from_cert.grade == self.certificates.get(
+            self.course_run2.edx_course_key).data.get('grade')
+
+    def test_compute_grade_for_non_fa(self):
+        """
+        Tests for _compute_grade_for_non_fa function
+        """
+        run3_data = self.user_edx_data.get_run_data(self.course_run3.edx_course_key)
+        run4_data = self.user_edx_data.get_run_data(self.course_run4.edx_course_key)
+
+        grade3 = api._compute_grade_for_non_fa(run3_data)
+        grade4 = api._compute_grade_for_non_fa(run4_data)
+
+        assert isinstance(grade3, api.UserFinalGrade)
+        assert isinstance(grade4, api.UserFinalGrade)
+
+        assert grade3.passed is False
+        assert grade3.grade == self.current_grades.get(
+            self.course_run3.edx_course_key).data.get('percent')
+        assert grade4.passed is True
+        assert grade4.grade == self.certificates.get(
+            self.course_run4.edx_course_key).data.get('grade')
+
+    def test_get_compute_func(self):
+        """
+        tests for _get_compute_func function
+        """
+        assert api._get_compute_func(self.course_run1) == api._compute_grade_for_fa
+        assert api._get_compute_func(self.course_run3) == api._compute_grade_for_non_fa
+
+    @patch('dashboard.api_edx_cache.CachedEdxDataApi.update_cached_current_grades', new_callable=MagicMock)
+    @patch('dashboard.api_edx_cache.CachedEdxDataApi.update_cached_certificates', new_callable=MagicMock)
+    @patch('dashboard.api_edx_cache.CachedEdxDataApi.update_cached_enrollments', new_callable=MagicMock)
+    @patch('backends.utils.refresh_user_token', autospec=True)
+    def test_refresh_cache_final_grade(self, refr, enr, cert, grades):
+        """
+        tests for _refresh_cache_final_grade function
+        """
+        api._refresh_cache_final_grade(self.user)
+        refr.assert_called_once_with(self.user.social_auth.get(provider=EdxOrgOAuth2.name))
+        enr.assert_not_called()
+        cert.assert_called_once_with(self.user, ANY)
+        grades.assert_called_once_with(self.user, ANY)
+
+    @patch('grades.api._compute_grade_for_non_fa', autospec=True)
+    @patch('grades.api._compute_grade_for_fa', autospec=True)
+    def test_get_final_grade(self, fa, non_fa):
+        """
+        tests for get_final_grade function
+        """
+        cruns = (self.course_run1, self.course_run3,)
+
+        for crun in cruns:
+            api.get_final_grade(self.user, crun)
+
+        assert fa.called is True
+        assert fa.call_count == 1
+        call_arg = fa.call_args_list[0][0][0]
+        assert isinstance(call_arg, UserCachedRunData)
+        assert call_arg.edx_course_key == self.course_run1.edx_course_key
+
+        assert non_fa.called is True
+        assert non_fa.call_count == 1
+        call_arg = non_fa.call_args_list[0][0][0]
+        assert isinstance(call_arg, UserCachedRunData)
+        assert call_arg.edx_course_key == self.course_run3.edx_course_key
```

Comment:
I think there is misunderstanding on the testing approach: I am testing a specific logic to finalize the grades and I think we need to go in the details here and not just see what is the final result.
I am leaving the tests as they are.